lib/rules/switch-final-break.js

```diff
@@ -0,0 +1,157 @@
+ * @author Eran Shabi <https://github.com/eranshabi>
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const lodash = require("lodash");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description:
+                "disallow an unnecessary `break` at the final clause of a switch statement.",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/switch-final-break"
+        },
+        fixable: "code",
+        messages: {
+            switchFinalBreakNever:
+                "Final clause in 'switch' statement should not end with 'break;'.",
+            switchFinalBreakAlways:
+                "Final clause in 'switch' statement should end with 'break;'."
+        },
+        schema: [
+            {
+                enum: ["never", "always"]
+            }
+        ],
+        type: "suggestion"
+    },
+    create(context) {
+        const requireBreak = context.options[0] ? context.options[0] : "never";
+
+        /**
+         * Returns the statements list of a given switch clause
+         * @param {ASTNode} node switch clause node
+         * @returns {Array} - the statements list of the switch clause
+         */
+        function getStatements(node) {
+            if (
+                node.consequent.length === 1 &&
+                node.consequent[0].type === "BlockStatement"
+            ) {
+
+                // block statement
+                const block = node.consequent[0];
+
+                return block.body;
+            }
+            return node.consequent;
+        }
+
+        /**
+         * Returns the last statement of a given switch clause
+         * @param {ASTNode} node switch clause node
+         * @returns {ASTNode} - the final statement of the switch clause
+         */
+        function getLastStatement(node) {
+            return lodash.last(getStatements(node));
+        }
+
+        return {
+            SwitchStatement(node) {
+                if (node.cases.length !== 0) {
+                    const lastCase = node.cases[node.cases.length - 1];
+                    const lastStatement = getLastStatement(lastCase);
+
+                    if (requireBreak === "always") {
+                        if (
+                            !lastStatement ||
+                            (lastStatement.type !== "BreakStatement" &&
+                                 getStatements(lastCase).every(statement => statement.type !== "ThrowStatement" && statement.type !== "ReturnStatement"))) {
+                            context.report({
```

Comment:
I'm not sure if this rule can be implemented without [code path analysis](https://github.com/eslint/eslint/blob/master/docs/developer-guide/code-path-analysis.md).

For example, `always` shouldn't report an error and shouldn't append a `break` statement to the following `case`:

```js
switch (foo) {
  case bar:
    if (baz) {
      return 1;
    } else {
      return 2;
    }
}
```

This example above should be correct, regardless of the option. A `break` statement at the end would be an unreachable code, as in [this demo](https://eslint.org/demo#eyJ0ZXh0Ijoic3dpdGNoIChmb28pIHtcbiAgY2FzZSBiYXI6XG4gICAgaWYgKGJheikge1xuICAgICAgcmV0dXJuIDE7XG4gICAgfSBlbHNlIHtcbiAgICAgIHJldHVybiAyO1xuICAgIH1cbiAgICBicmVhaztcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsiZ2xvYmFsUmV0dXJuIjp0cnVlfX0sInJ1bGVzIjp7Im5vLXVucmVhY2hhYmxlIjoyfSwiZW52Ijp7fX19).

I think that the current implementation doesn't handle cases like this well?

`no-fallthrough` rule might be a very good example of the code path analysis for this case.